docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,265 @@
+# Sync improvements
+
+* Author(s): Cornelius Weig (@corneliusweig)
+* Design Shepherd: Techal Desai (@tejal29)
+* Date: 03/20/2019
+* Status: Draft
+
+## Background
+
+Currently, skaffold config supports `artifact.sync` as a way to sync files
+directly to pods. So far, artifact sync requires a specification of sync
+patterns like
+
+```yaml
+sync:
+  '*.js': app/
+```
+
+This is error prone and unnecessarily hard to use, because the destination is
+already contained in the Dockerfile for docker build. (see #1166, #1581).
+In addition, the syncing needs to handle special cases for globbing and often
+requires a long list of sync patterns (#1807).
+
+Furthermore, builders should be able to inform Skaffold about additional sync paths (#1704).
+This will result in much faster deploy cycles, without users having to bother about these optimizations.
+
+## Design
+
+### Goals
+The sync mechanism in skaffold should be able to infer the destination location of local files.
+
+For example: given the following Dockerfile
+
+```dockerfile
+FROM nginx
+WORKDIR /var
+COPY static *.html www/
+ADD nginx.conf .
+```
+
+The sync locations should be inferred as
```

Comment:
The sync works for this example as it's a straightforward copy.  But as soon as we involve some transformation tool (e.g., webpack or babel or tsc), then this no longer works.

Reply:
I think for webpack/tsc the sync would still do the right thing, as long as webpack/tsc runs inside the container (see #1826).

However, you are right that this
```dockerfile
FROM golang as builder
COPY main.go .
RUN go build

FROM scratch
COPY --from=builder main .
```
will not work as intended.

However, I think that there can be no general solution here. In the end, only the developers know what files are static and what needs to be processed. This is what the sync rules are for after all.

Nevertheless, we could think about a warn/info if typical static files cause a rebuild. For example, if somebody changes a `index.html` file and no sync rule is set up, we could advise to do so. But I have to say, I'm not a big fan of this.

Reply:
When I read through the earlier version, it wasn't clear that your approach required the user to opt-in to the change by defining a sync-rule in the skaffold.yaml.  Could you please call this out more explicitly above?